File: javascript/wordy/wordy.js
// REVIEW: Check the comments on exercism.io and see if implementing the capturing groups
// regex thing makes things cleaner and faster.
exports.WordProblem = WordProblem;
exports.ArgumentError = ArgumentError;

function ArgumentError() {}

var operations = {
	plus: function(lval, rval) {
		return lval + rval;
	},
	
	minus: function(lval, rval) {
		return lval - rval;
	},
	
	multiplied: function(lval, rval) {
		return lval * rval;
	},
	
	divided: function(lval, rval) {
		return lval / rval;
	}
};

function WordProblem(problem) {
	this.question = problem;
}

WordProblem.prototype.isValidQuestion = function() {
	wellFormedrx = /What\ is(\ -?[0-9]*\ (plus|minus|divided\ by|multiplied\ by))*\ -?[0-9]*\?/;
	return wellFormedrx.test(this.question);
};

WordProblem.prototype.answer = function() {
	if (!this.isValidQuestion(this.question)) {
		throw new ArgumentError();
	}
	
	// TODO: There must be a better way than using replace.
	var cInput = this.question.replace(/\?/, '').replace(/\ by/g, '').split(' ');

	var result = 0;
	var limit = cInput.length;
	for (var i = 0; i < limit; ++i) {
		if (operations[cInput[i]]) {
			if (result) {
				result = operations[cInput[i]](parseInt(result), parseInt(cInput[i + 1]));
			} else {
				result = operations[cInput[i]](parseInt(cInput[i - 1]), parseInt(cInput[i + 1]));
			}
		}
	}
	
	return result;
};

File: python/wordy/wordy.py
import re


operations = {
    'plus': (lambda lval, rval: lval + rval),
    'minus': (lambda lval, rval: lval - rval),
    'multiplied': (lambda lval, rval: lval * rval),
    'divided': (lambda lval, rval: lval / rval)
}

def valid_question(question):
    well_formed_rx = r"What\ is(\ -?[0-9]*\ (plus|minus|divided\ by|multiplied\ by))*\ -?[0-9]*\?"
    return re.match(well_formed_rx, question)

def calculate(question):
    if not valid_question(question):
        raise ValueError("Invalid question.")

    clean_input = re.sub(r"\ by", '', re.sub(r"\?", '', question)).split(' ')

    result = 0
    limit = len(clean_input)
    for i in range(limit):
        if clean_input[i] in operations:
            if result:
                result = operations[clean_input[i]](int(result), int(clean_input[i + 1]))
            else:
                result = operations[clean_input[i]](int(clean_input[i - 1]), int(clean_input[i + 1]))

    return result